Add PublicKey::to_sort_key method for use with sorting by key

junderw opened the pull request:
Replaces #524 

See previous PR for reasoning.

This solution is a little more straightforward. The name and documentation should be enough to prevent misuse.

We can also impl a to_sort_key for any CompressedKey added later. (or just impl Ord in a BIP67 compliant way)

TODO:

- [x] Add more sorting test vectors. Ideas of edge cases to test are welcome.

junderw commented:
I added a vector with mixed uncompressed and compressed.

Other vectors are welcome.

junderw commented on src/util/key.rs:
````diff
@@ -182,6 +182,51 @@ impl PublicKey {
+    ///
+    /// assert_eq!(unsorted, sorted);
+    /// ```
+    pub fn get_sort_key(&self) -> [u8; 65] {
````
Sounds good.

Kixunil commented on src/util/key.rs:
```diff
@@ -182,6 +182,51 @@ impl PublicKey {
+    /// `*_by_key` methods on slice. Pass the method into the sort method directly.
+    /// (ie. `PublicKey::get_sort_key`)
+    ///
+    /// If every `PublicKey` in the slice is `compressed == true` then this will sort
```
> As arrays are checked in order until the first non-equal value which returns early.

Oh, yes, forgot about this. :sweat_smile:
Moving around those zeroes may still be a bit costly, not sure. Maybe not enough to maintain another function.

I'm surprised the optimizer doesn't seem to be able to optimize-out the comparison. I though it does and [tried to prevent it](https://play.rust-lang.org/?version=nightly&mode=release&edition=2021&gist=da4f3a503d753a8c1eede283c5949e53) but it seems to be same.

junderw replied:
I have changed this part of the code. Resolving.

junderw commented:
New fixes:

1. `1.41.1` CI was broken because const generics don't exist, which means [u8; 65] is not Ord. This is fixed by converting to a tuple of `(u8, [u8; 32], Option<[u8; 32]>)` which will make the same order. (Only compressed keys will have a None in the last position)
2. Changed the name to `to_sort_key` which is in line with the [API Guidelines](https://rust-lang.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv)
3. Added a key to the last vector which mirrors another key on the X axis. This creates a sort with the first 33 bytes exactly the same and only the Y value differs.

Kixunil commented:
Hmm, `Option` adds one more byte to the type and given your reasoning it shouldn't be needed. Either way the type is becoming ugly so I propose to either:

* Make it opaque; we can change it to type alias one day when we bump MSRV again
* Throw it away and change `to_bytes` to return `SerializedPublicKey` exactly as I wrote it. (You reminded me `Ord`/`PartialOrd` is missing, will add soon.)

I'm starting to like the latter because it serves another purpose too, so it's more useful but the former may be very slightly faster when sorting (skips computing length; though length computing is very fast)

Kixunil reviewed (commented):
I'm open to accepting this (modulo doc) but would like to see more discussion from others, especially regarding idea of changing `to_bytes`.

junderw commented:
I improved the doc example and squashed everything into 1 commit.

apoelstra commented:
I'm a little tempted to bikeshed on the `Option` thing ... yes, we save a byte, but we plausibly speed up comparison time.

On the other hand I really doubt the speedup is measurable, so without benchmarks I won't push it. I'm just mentioning it for the record.

junderw commented:
re: Option: I don't care either way. I ran some simple benchmarks locally (on rust playground) and it seems like an extra 32 0 == 0 u8 comparisons are about 0.8 ns.

However, like I said in the discussion, if you think about the most common use cases for this sorting, having two or more compressed pubkeys that are equal (which is the condition for hitting the 0 comparisons) is always a sign of something going wrong.

I can't think of one use case for sorting PublicKey objects where having two identical compressed keys isn't an indication of a serious bug somewhere else, and most correct programs will therefore never even hit the condition for the extra compares.

I would be interested if someone can think of a use case where two equal pubkeys isn't a sign of another issue...

But like you said, I think we can just merge as-is.

Just wanted to share my self-justification.

junderw commented:
Also as an added note for reference on how Bitcoin Core does it:

[Here](https://github.com/bitcoin/bitcoin/blob/v23.0/src/script/descriptor.cpp#L785-L803) is the area where sortedmulti() logic is written. It just uses std::sort on a vector of CPubKey.

[Here](https://github.com/bitcoin/bitcoin/blob/v23.0/src/pubkey.h#L118-L137) is the logic for CPubKey comparison operators. vch is a 65 byte array with the first byte being 0x02|0x03|0x04 (so same format as we have here) and being compared... then memcmp is used to compare the rest (size() returns 33 when compressed, so it won't compare the 32 empty bytes in the array)

tcharding reviewed (approved):
ACK 24f0441d54ac91066ea492916158593d992f1b11

Kixunil commented:
@junderw the Core logic looks very similar to my suggestion. :)

Kixunil commented on src/util/key.rs:
```diff
@@ -182,6 +182,72 @@ impl PublicKey {
+    /// `SortKey` is not too useful by itself, but it can be used to sort a
+    /// `[PublicKey]` slice using `sort_unstable_by_key`, `sort_by_cached_key`,
+    /// `sort_by_key`, or any of the other `*_by_key` methods on slice.
+    /// Pass the method into the sort method directly. (ie. `PublicKey::to_sort_key`)
```
I think it'd be nice to document that the conversion is non-trivial, so cached versions may perform better.

Kixunil reviewed (commented):
I'm willing to accept this code if others say "we don't want to have `SerializedPublicKey`". If others want `SerializedPublicKey` then this would be duplication of code since that one would provide same thing with additional features.

However if others think that `to_sort_key` returning type alias of `SerializedPublicKey` is good for consumer code clarity, I'm also willing to accept this. :)

apoelstra commented:
I kinda like the idea of having `SerializedPublicKey` actually ... it would make a lot of Core behavior easy to emulate, not just sort order.

junderw commented:
Since the link wasn't posted here I'll post a link to the [`SerializedPublicKey`](https://github.com/BP-WG/bp/blob/fcae98363935f3f410cf237e7e968e25f078383a/bitcoin_keys/src/ecdsa/serialized_public_key.rs) that was posted in the other thread.

junderw commented:
tbh I think it would be better if SerializedPublicKey Derefed into bitcoin::PublicKey.

Since we already would be manually impl all the Ord Eq etc. we might as well hold on to the bitcoin::PublicKey internally.

junderw commented:
Then it would be:

1. All functionality is on bitcoin::PublicKey
2. If the serialization is important, .into() will consume and wrap the bitcoin::PublicKey in a SerializedPublicKey with the 65 byte array.

It will increase the data size of each key, but makes it more flexible, and we don't need to reimplement anything / worry about conversions when dealing with secp256k1 methods.

junderw commented:
Looking into it, it seems like most functionality on bitcoin::PublicKey serializes internally, so a simple Deref would leave us with an awkward inefficient API where SPK would be caching the serialization, but using one of the Derefed methods would serialize again and not make use of the cached serialization in SPK.

Kixunil commented:
> I think it would be better if SerializedPublicKey Derefed into bitcoin::PublicKey.

Even if I agreed, it's impossible without ~doubling it's size. `SerializedPublicKey` is just a bunch of bytes and it's equivalent to `SmallVec<const LEN = 65, u8>`

Kixunil commented:
Hmm, but maybe methods for computing hashes of `SerializedPublicKey` would be helpful. And maybe also `CachedPublicKey` as a new type.

junderw commented:
I think everyone's vision of what SerializedPublicKey should be is all over the place and causing confusion.

Currently, bitcoin::PublicKey can:

1. Create an instance from secp256k1::Pubkey which is guaranteed valid key.
2. Get keyhash and wkeyhash (uses serialize*())
3. Write into writer / output a Vec<u8> (uses serialize*())
4. Read from reader / instanciate from &[u8] (uses the fallible secp256k1::Pubkey::from_slice)

So SerializedPublicKey should:

1. Use secp256k1::Pubkey::from_slice etc. for any instantiation to ensure valid point / accept secp256k1::Pubkey
2. Get keyhash and wkeyhash
3. Write into writer
4. Read from reader (get an instance)

In which case... we'd basically be replacing PublicKey completely...

Maybe the `SerializedPublicKey` discussion should happen separately and be introduced along with a deprecation of PublicKey... or replace PublicKey with it and release as breaking change.

SortKey and to_sort_key seems like a minimal non-breaking addition that provides a simple use case that isn't too performance critical to begin with.

Perhaps we should just merge and continue discussion about the specifics of SerializedPublicKey for a breaking change?

Thoughts?

apoelstra commented:
If we validate `SerializedPublicKey` then we ought to just go all the way and convert it to a `PublicKey`. This is barely more expensive and would save space.

Serializing is basically free. Parsing and validation are not. If we were to have a `SerializedPublicKey` the point would be to basically be a wrapper around a bytestring, analogous to `CPublicKey` in Core.

junderw commented:
So basically, what @apoelstra wants is something similar to what @Kixunil except without using `secp256k1::PublicKey` at all.

Basically any instance would not be validated upon instantiation beyond a simple length check and maybe check that first byte is in the set [2,3,4] (Bitcoin Core also allows [6,7] which I am not sure of their meaning beyond the [2,3] bit patterns with a 4 bit added on)

There is a public method on CPubKey that does simple validation (which checks that it isn't the `Invalid` default constructor... not sure if this is needed for Rust) and then a more expensive validation.

It also looks like they add overrides to allow CPubKey to be used similar to its underlying data structure, so we should implement Deref<Target = [u8]> and Index<usize> and IndexMut<usize>

We might as well just name it CPubKey, I guess.

However, I wonder about the utility of some of the methods for this library. (some of them seem to be working around C++ limitations)

Kixunil commented:
Firstly, we're programming in Rust, not C++ so please avoid taking too much inspiration from Core design.

I'm not sure if we should support directly converting arbitrary slices to `SerializedPublicKey`. I'd prefer to avoid it and only implement it i someone has an actual use case.

> Bitcoin Core also allows [6,7] which I am not sure of their meaning

Those are exotic encodings we don't support here.

> we should implement Deref<Target = [u8]> and Index< usize >

The code I reference indeed has `Deref` but not `Index` because that one is provided by deref coercions.

> We might as well just name it CPubKey, I guess.

Please no. This is C++/Core naming convention ("C" meaning "class") and it could be confusing since we will not match it entirely. I'm not opposed to saying in the doc "This is similar, but *not* exactly same as `CPubKey` in core." though.

> However, I wonder about the utility of some of the methods on Bitcoin Core's CPubKey for the port to this library. (some of them seem to be working around C++ limitations)

Exactly, I think we should stick to it being a heap-allocation-free replacement for `Vec<u8>` returned from `to_bytes()`, maybe with hashing methods.

junderw commented:
I'm just trying to figure out what it is about the CPubKey design that we want to emulate besides the ordering.

It (Core's CPubKey) validates signatures, verifies signatures, checks if signatures are lowS, recovers pubkeys from signatures+hashes into itself, derives child keys using bip32 by passing chain codes as an argument etc... all of which are... not things I would normally expect we'd want to emulate.

My stance is still unchanged:

- This should be merged as-is to at least give a method for ordering pubkeys, and it's a simple opaque type.
- Discussion on SerializedPublicKey and its API should be in a different issue / PR (since it is a source of endless bike-shedding on the nitty gritty details)
  - I think there is a high chance that whatever SerializedPublicKey eventually turns into will end up completely replacing bitcoin::PublicKey, as well.

Kixunil commented:
OK, I take back my willingness to mention `CPubKey` in the docs, it's completely confusing and it shouldn't affect the design at all. I'm taking Rust-first approach. I only want `SerializedPublicKey` to be a bunch of bytes similar to `Box<[u8]>` (yep, not even `Vec<u8>`) with the only difference being it's not heap allocated. The fact that it happens to provide ordering that happens to be in line with Core/BIP-67 is a nice side effect.

> I think there is a high chance that whatever SerializedPublicKey eventually turns into will end up completely replacing bitcoin::PublicKey

I think the chance is near zero. It has concept NACK from me and I'd need some very serious argument to consider it.

sanket1729 reviewed (commented):
post merge ACk 24f0441d54ac91066ea492916158593d992f1b11
